proposals/self-types.md

```diff
@@ -0,0 +1,460 @@
+# Self types
+
+* **Type**: Design proposal
+* **Authors**: Maksim Grankin
+* **Status**: Prototype implemented
+* **Issue**: [KT-6494](https://youtrack.jetbrains.com/issue/KT-6494)
+
+
+A **Self type** is type that refers to the receiver type.
+
+```kotlin
+open class A {
+    fun foo(): Self {
+        return this;
+    }
+}
+
+class B : A {
+    
+}
+
+val x: B = B().foo(); // foo return type is B
```

Comment:
val x:B = B().foo().bar()

Reply:
`;` will be removed, for `bar` function take a look at the previous comment.

Thanks!